src/preamble.js

```diff
@@ -164,42 +170,42 @@ function ccall (ident, returnType, argTypes, args, opts) {
     }
   }
   var ret = func.apply(null, cArgs);
-#if ASSERTIONS
 #if EMTERPRETIFY_ASYNC
-  if ((!opts || !opts.async) && typeof EmterpreterAsync === 'object') {
-    assert(!EmterpreterAsync.state, 'cannot start async op with normal JS calling ccall');
+  if (typeof EmterpreterAsync === 'object' && EmterpreterAsync.state) {
```

Comment:
the check for `.state` could maybe be pulled out? Something like: if opts.async, then assert on state != 0, and otherwise, assert on state == 0?

Reply:
We already discussed this before in https://github.com/kripken/emscripten/pull/6586#discussion_r191964769, we can't assert on state != 0 if opts.async is true.

Reply:
Sorry, then I think I didn't understand the rest of that discussion.

I thought that we require the user to explicitly say `async: true` to do an async call, and I don't see how it is not an error to not actually do an async call when that flag is specified. What am I missing.

Reply:
You can also turn it around and ask, why should it be an error?

Whether an operation actually does an async operation or not is an implementation detail.
In some cases, you call a function and don't know yet if it will do an async operation or not.
To give an example: Imagine a function that returns some resource. The resource could be cached, and in this case returned immediately (no async operation) or it might need to be fetched from the network (async).

Reply:
On that higher level, I agree - when requesting a resource, logically that operation may need to be async or not. But I disagree that `ccall` should abstract over those two - it's a low-level API, and should be as simple as possible. An API returning either a concrete result or a promise is fairly complex, in my opinion - it would require the callers to all handle both cases, in particular. So at the `ccall` level, I think we should be clear: an operation is either async, or it is not. People can then build higher-level abstractions over that if it is useful.

Reply:
My implementation [wraps the result in a promise](https://github.com/kripken/emscripten/pull/6586/files#diff-5df1ca8b00952a4230cf0be8d8bf3682R191) when the operation return synchronously, so you always get a promise back when calling with `{async: true}`.

Reply:
I see, thanks. Let me do some more reading here.